README.md
# RULA
 This is an automatic implementation of the Rapid Upper Limb Assessment by using data from IMU sensors

Run app-dev.py to start the App


changelog.txt
2018-05-04
+ batch process
+ file output
+ app won't close
+ changed leg_score question
+ bug fixes: body rotation, body tilt, head rotation, head tilt
- right upper_body Bug (reason unknown)


library/app.py
# coding: utf-8
#!/usr/bin/env python
"""
combines the core functions
"""
import numpy as np
import time
import datetime as dt
import sys
import glob
import get_data as gd
import rula_scores as rs


def app_rula(path):
    #path = "../data/rula.csv
    print("===================\n===================")
    print("Starting...")
    print("Importing Data...{%s}" %path)
    # generating the single Scores
    info, header, data = gd.get_data(path)
    upperArm_left, upperArm_right = rs.upperArm_score(header, data)
    lowerArm_left, lowerArm_right = rs.lowerArm_score(header, data)
    wrist_left, wrist_right, suppination_left, suppination_right = rs.wrist_score(header, data)
    neck = rs.neck_score(header, data)
    upperBody = rs.upperBody_score(header, data)
    leg = rs.leg_score()

    # generating the combined scores fromt the RULA tables
    if upperArm_left.shape == upperArm_right.shape == lowerArm_left.shape == lowerArm_right.shape == wrist_left.shape == wrist_right.shape == suppination_left.shape == suppination_right.shape == neck.shape == upperBody.shape:
        final_array = np.zeros((len(upperArm_left), 7)) #[A[l,r,c],B,C[l,r,c]]
        print("Looping for Scores: ")
        # Progressbar
        toolbar_width = 80
        # setup progressbar
        sys.stdout.write("[%s]" % (" " * toolbar_width))
        sys.stdout.flush()
        sys.stdout.write("\b" * (toolbar_width + 1)) # return to start of line, after '['
        j = 0
        for i in range(len(final_array)):
            j += 1
            if j == int(len(final_array)/toolbar_width):
                j = 0
                sys.stdout.write("#")
                sys.stdout.flush()

            # scores from RULA tables
            final_array[i, 0] = rs.score_table_a(upperArm_left[i], lowerArm_left[i], wrist_left[i], suppination_left[i], i)
            final_array[i, 1] = rs.score_table_a(upperArm_right[i], lowerArm_right[i], wrist_right[i], suppination_right[i], i)
            final_array[i, 2] = np.mean([final_array[i, 0], final_array[i, 1]])
            final_array[i, 3] = rs.score_table_b(neck[i,0], upperBody[i,0], leg)
            final_array[i, 4] = rs.score_table_c(final_array[i, 0], final_array[i, 3])
            final_array[i, 5] = rs.score_table_c(final_array[i, 1], final_array[i, 3])
            final_array[i, 6] = rs.score_table_c(final_array[i, 2], final_array[i, 3])

    else:
        print("inconsistent shapes of matrices")
        print(upperArm_left.shape)
        print(upperArm_right.shape)
        print(lowerArm_left.shape)
        print(lowerArm_right.shape)
        print(wrist_left.shape)
        print(wrist_right.shape)
        print(suppination_left.shape)
        print(suppination_right.shape)
        print(neck.shape)
        print(upperBody.shape)

    output =("""
=========================================================
=========================================================
FINAL MEAN VALUES OF RULA TABLES)
A-left\tA-right\tA-comb\tB\tC-left\tC-right\tC-comb\n
%.2f\t%.2f\t%.2f\t%.2f\t%.2f\t%.2f\t%.2f
=========================================================
========================================================="""
    %(np.mean(final_array[:,0]), np.mean(final_array[:,1]), np.mean(final_array[:,2]),
    np.mean(final_array[:,3]), np.mean(final_array[:,4]), np.mean(final_array[:,5]),
    np.mean(final_array[:,6])))

    print(output)
    return output


def write_to_file(filename, output):
    """write results to *.tsv file
    file will be stored in original data folder
    filename = string, output = string
    """
    path = "../data/" + "scored_" + filename + ".tsv"
    fObj = open(path, "w+")
    fObj.write(output)
    fObj.close()


print("""#################################################\n
#                                               #\n
#                                               #\n
#    /////    |   |    |        /////           #\n
#    |   |    |   |    |        |   |           #\n
#    /////    |   |    |        /////           #\n
#    | /      |   |    |        |   |           #\n
#    |  /     /////    /////    |   |     v0.35 #\n
#                                               #\n
#                                               #\n
#################################################
""")
start_time = dt.datetime.now()

mode = input("""Choose Mode:\n
Please make sure all data-files are within the data folder\n
For Batch-processing Mode press b\n
For Single-File Processing press s:\t""")

if mode == "b":
    file_list = glob.glob("../data/*.csv")
    print("%i files were found" %len(file_list))
    for i,item in enumerate(file_list):
        print("\nAnalyzing file %i out of %i\n" %(i, len(file_list)))
        output = app_rula(item)
        f_name = item.split('\\')[-1]
        write_to_file(f_name.split('.')[0], output)

    print("Process finished. Exiting.")

elif mode == "s":
    while True:
        try:
            fObj = input("Please enter filename (without csv-extension): \n\t")
            path = "../data/" + fObj + ".csv"
            print(gd.get_data.__doc__)
            output = app_rula(path)
            write_to_file(fObj, output)
            input("To Exit, press CTRL+C\nTo keep going, press ENTER\n")
        except KeyboardInterrupt:
            sys.exit(0)
        except:
            print("Unexpected error:", sys.exc_info()[0])
            print("Something went wrong. Try again!")
            raise

end_time = dt.datetime.now()
time_delta = end_time - start_time
print("Process Duration: " time_delta.total_seconds())


library/app-dev.py
# coding: utf-8
#!/usr/bin/env python
"""
combines the core functions
"""
import numpy as np
import time
import datetime as dt
import sys
import glob
import get_data as gd
import rula_scores as rs
import report as rp


def app_rula(path):
    #path = "../data/rula.csv
    print("===================\n===================")
    print("Starting...")
    print("Importing Data...{%s}" %path)
    # generating the single Scores
    info, header, data = gd.get_data(path)
    upperArm_left, upperArm_right = rs.upperArm_score(header, data)
    lowerArm_left, lowerArm_right = rs.lowerArm_score(header, data)
    wrist_left, wrist_right, suppination_left, suppination_right = rs.wrist_score(header, data)
    neck = rs.neck_score(header, data)
    upperBody = rs.upperBody_score(header, data)
    leg = rs.leg_score()

    # generating the combined scores fromt the RULA tables
    if upperArm_left.shape == upperArm_right.shape == lowerArm_left.shape == lowerArm_right.shape == wrist_left.shape == wrist_right.shape == suppination_left.shape == suppination_right.shape == neck.shape == upperBody.shape:
        final_array = np.zeros((len(upperArm_left), 7)) #[A[l,r,c],B,C[l,r,c]]
        print("Looping for Scores: ")
        # Progressbar
        toolbar_width = 80
        # setup progressbar
        sys.stdout.write("[%s]" % (" " * toolbar_width))
        sys.stdout.flush()
        sys.stdout.write("\b" * (toolbar_width + 1)) # return to start of line, after '['
        j = 0
        for i in range(len(final_array)):
            j += 1
            if j == int(len(final_array)/toolbar_width):
                j = 0
                sys.stdout.write("#")
                sys.stdout.flush()

            # scores from RULA tables
            final_array[i, 0] = rs.score_table_a(upperArm_left[i], lowerArm_left[i], wrist_left[i], suppination_left[i], i)
            final_array[i, 1] = rs.score_table_a(upperArm_right[i], lowerArm_right[i], wrist_right[i], suppination_right[i], i)
            final_array[i, 2] = np.mean([final_array[i, 0], final_array[i, 1]])
            final_array[i, 3] = rs.score_table_b(neck[i,0], upperBody[i,0], leg)
            final_array[i, 4] = rs.score_table_c(final_array[i, 0], final_array[i, 3])
            final_array[i, 5] = rs.score_table_c(final_array[i, 1], final_array[i, 3])
            final_array[i, 6] = rs.score_table_c(final_array[i, 2], final_array[i, 3])

    else:
        print("inconsistent shapes of matrices")
        print(upperArm_left.shape)
        print(upperArm_right.shape)
        print(lowerArm_left.shape)
        print(lowerArm_right.shape)
        print(wrist_left.shape)
        print(wrist_right.shape)
        print(suppination_left.shape)
        print(suppination_right.shape)
        print(neck.shape)
        print(upperBody.shape)

    output =("""
=========================================================
=========================================================
FINAL MEAN VALUES OF RULA TABLES)
A-left\tA-right\tA-comb\tB\tC-left\tC-right\tC-comb\n
%.2f\t%.2f\t%.2f\t%.2f\t%.2f\t%.2f\t%.2f
=========================================================
========================================================="""
    %(np.mean(final_array[:,0]), np.mean(final_array[:,1]), np.mean(final_array[:,2]),
    np.mean(final_array[:,3]), np.mean(final_array[:,4]), np.mean(final_array[:,5]),
    np.mean(final_array[:,6])))

    print(output)
    return output, final_array


def write_to_file(filename, output):
    """write results to *.tsv file
    file will be stored in original data folder
    filename = string, output = string
    """
    path = "../data/" + "scored_" + filename + ".tsv"
    fObj = open(path, "w+")
    fObj.write(output)
    fObj.close()


print("""#################################################\n
#                                               #\n
#                                               #\n
#    /////    |   |    |        /////           #\n
#    |   |    |   |    |        |   |           #\n
#    /////    |   |    |        /////           #\n
#    | /      |   |    |        |   |           #\n
#    |  /     /////    /////    |   |     v0.30 #\n
#                                               #\n
#                                               #\n
#################################################
""")
start_time = dt.datetime.now()

mode = input("""Choose Mode:\n
Please make sure all data-files are within the data folder\n
For Batch-processing Mode press b\n
For Single-File Processing press s:\t""")

if mode == "b":
    file_list = glob.glob("../data/*.csv")
    print("%i files were found" %len(file_list))
    for i,item in enumerate(file_list):
        print("\nAnalyzing file %i out of %i\n" %(i, len(file_list)))
        output, final_array = app_rula(item)
        f_name = item.split('\\')[-1]
        write_to_file(f_name.split('.')[0], output)

    print("Process finished. Exiting.")

elif mode == "s":
    while True:
        try:
            fObj = input("Please enter filename (without csv-extension): \n\t")
            path = "../data/" + fObj + ".csv"
            print(gd.get_data.__doc__)
            output, final_array = app_rula(path)
            write_to_file(fObj, output)

            rp.rula_to_pdf("test", final_array)
            end_time = dt.datetime.now()
            time_delta = end_time - start_time
            print("Process Duration: %1.4f" %time_delta.total_seconds())

            input("To Exit, press CTRL+C\nTo keep going, press ENTER\n")
        except KeyboardInterrupt:
            sys.exit(0)
        except:
            print("Unexpected error:", sys.exc_info()[0])
            print("Something went wrong. Try again!")
            raise


library/get_data.py

# coding: utf-8
#!/usr/bin/env python
"""
module for importing data and analyzing due to rula
"""

import numpy as np
import csv
import re


def get_data(fID):
    """Function ro read MR3 and QTM files

    input needs to be *.csv or *.tsv. Required decimal separator is "."
    Returns file information (dict), header information (dict) and data (np.array)
    """
    info = dict()
    with open(fID, encoding='utf-8') as csvfile:
        print(fID)
        try:
            if re.search(".csv", fID):
                delimiter = ";"
                csvcontent = csv.reader(csvfile, delimiter=delimiter)
                for a,row in enumerate(csvcontent):
                    if row[0] == "Name":
                        info[row[0]] = row[1]
                    elif row[0] == "Frequency":
                        info[row[0]] = row[1]
                    elif row[0] == "Date":
                        info[row[0]] = row[1]
                    elif re.search("Time", row[0]):
                        header = np.array(row)
                        break

            elif re.search(".tsv", fID):
                delimiter = "\t"
                csvcontent = csv.reader(csvfile, delimiter=delimiter)
                for a,row in enumerate(csvcontent):
                    if row[0] == "Frame":
                        header = np.array(row)
                        a += 1
                        break
                    else:
                        info[row[0]] = row[1::]

            data = np.genfromtxt(fID, delimiter=delimiter, skip_header=a+1, dtype=np.float64)
        except:
            print("could not load data. Loading capturing information failed. Please check file format.")


    print(info)
    return info, header, data



def read_all_files(fID):
    """Generates a list of paths to files

    inuput is ASCII file; file must contain paths to other files
    """
    file_list = []
    with open(fID, encoding='utf-8') as current_file:
        for line in current_file:
            line = line.replace('\n', '')
            file_list.append(line)

    return file_list



def moving_average_1D(data, window):
    """smooths data according to moving average algorithm

    input should be array or list; returns 1-D array
    """
    smoothed_array = np.zeros(len(data)-window)
    for i in range(len(smoothed_array)):
        smoothed_array[i] = np.mean(data[i, i+window])

    return smoothed_array


library/rectangle.py
import matplotlib.pyplot as plt
import matplotlib.patches as pt
import numpy as np
import datetime
from matplotlib.backends.backend_pdf import PdfPages

plt.style.use('ggplot')
# Create the PdfPages object to which we will save the pages:
# The with statement makes sure that the PdfPages object is closed properly at
# the end of the block, even if an Exception occurs.
x = np.linspace(-50, 50, 100)
y1 = np.sin(x)
y2 = np.cos(x)
y3 = np.power(x,3)
y4 = np.sqrt(np.abs(x))

width, height = 25/2.54, 29/2.54
with PdfPages('multipage_pdf.pdf') as pdf:
    fig = plt.figure(figsize=(width, height))
    a = plt.axes([.05, .8, .4, .16])
    plt.plot(x, y1)
    plt.tick_params(axis="both", labelsize=0.0)
    a = plt.axes([.55, .8, .4, .16])
    plt.tick_params(axis="both", labelsize=0.0)
    plt.plot(x, y2)
    a = plt.axes([.05, .6, .4, .16])
    plt.tick_params(axis="both", labelsize=0.1)
    plt.plot(x, y3)
    a = plt.axes([.55, .6, .4, .16])
    plt.tick_params(axis="both", labelsize=0.1)
    plt.axvspan(-10, 10, alpha=0.5)
    plt.plot(x, y4)

    # plt.rc('text', usetex=True)
    # rect1 = pt.Rectangle((0, 0), 10, 10)
    # ax1.add_patch(rect1)
    # ax1.set_title('Plot Two')
    # pdf.attach_note("plot of sin(x)")  # you can add a pdf note to
    #
    #
    # plt.rc('text', usetex=False)
    # rect2 = pt.Rectangle((0, 0), 10, 10)
    # ax2.add_patch(rect2)
    # ax2.set_title('Plot Three')

    pdf.savefig()  # or you can pass a Figure object to pdf.savefig
    plt.close()

    # We can also set the file's metadata via the PdfPages object:
    d = pdf.infodict()
    d['Title'] = 'Multipage PDF Example'
    d['Author'] = u'Jouni K. Sepp\xe4nen'
    d['Subject'] = 'How to create a multipage pdf file and set its metadata'
    d['Keywords'] = 'PdfPages multipage keywords author title subject'
    d['CreationDate'] = datetime.datetime(2009, 11, 13)
    d['ModDate'] = datetime.datetime.today()


library/report.py
"""
This is a demo of creating a pdf file with several pages,
as well as adding metadata and annotations to pdf files.
"""

import datetime
import numpy as np
from matplotlib.backends.backend_pdf import PdfPages
import matplotlib.patches as pt
import matplotlib.pyplot as plt
import matplotlib.image as im

# Create the PdfPages object to which we will save the pages:
# The with statement makes sure that the PdfPages object is closed properly at
# the end of the block, even if an Exception occurs.
def single_plots_graph(l, b, w, h, cols, title, data):
    """This function sets the settings for the different plots using several parameters
    """
    subFig = plt.axes([l, b, w, h])
    plt.title("!!")
    for i in range(len(cols)):
        plt.plot(data[:, cols[i]]), plt.plot(data[:, cols[i]], color='black')

    plt.axhspan(0.5, 2.5, alpha=0.15, color="green")
    plt.axhspan(2.5, 6.5, alpha=0.15, color="yellow")
    plt.axhspan(6.5, 10, alpha=0.15, color="red")
    plt.yticks(np.arange(1, 10, step=1))
    plt.title(title, fontsize=8)
    plt.tick_params(axis="both", labelsize=5)


def single_plots_hist(l, b, w, h, index, title, data):
    """This function defines and plots the data as histograms
    """
    subFig = plt.axes([l, b, w, h])
    plt.hist(data[:,index], bins=18, range = (1,9), color='black', align='mid')
    plt.title(title, fontsize=8)
    plt.axvspan(1, 3, alpha=0.15, color="green")
    plt.axvspan(3, 7, alpha=0.15, color="yellow")
    plt.axvspan(7, 10, alpha=0.15, color="red")
    plt.tick_params(axis="both", labelsize=5)
    plt.yticks(np.arange(0,101,step=20))


def ampel_plot(data):
    """Perfoming an ample plot for final scores and each of the VALUES
    """
    l, b, w, h = .03, .03, .1, .1
    title = ["Arm li", "Arm re", "Arm combo", "Oberkörper", "Ober-Arm li", "Ober-Arm re", "Total"]
    for i in range(data.shape[1]):
        x, y = 0.5, 0.5
        if np.mean(data[:,i]) <= 3:
            color = "green"
        elif (np.mean(data[:,i]) > 3) and (np.mean(data[:,i]) <= 6):
            color = "yellow"
        else:
            color = "red"

        subPlot = plt.axes([l, b, w, h])
        circle = pt.Circle((x, y), radius=.45, color=color, fill=True)
        subPlot.add_artist(circle)
        plt.tick_params(axis="both", labelsize=0)
        plt.text(x, y, s=str(np.around(np.mean(data[:,i]), decimals=2)))
        plt.title(title[i], fontsize=8)
        l += 1/7


def rula_to_pdf(name, data):
    """This function generates a *.pdf Report
    name = Output Path (string); data = array (input)
    """
    plt.style.use('fivethirtyeight')
    width, height = 25/2.54, 29/2.54
    l, b, w, h = .95, .95, .95, .95
    image_path = ["../pics/skelett.png", "../pics/skelett_back.png"]
    with PdfPages(name + '.pdf') as pdf:
        fig = plt.figure(figsize=(width, height), dpi=200)
        # -1- first page
        # first half of page
        plt.axes([.02, .52, .99, .5])
        plt.text(0.4,0.9, s="Verlauf über die Zeit")
        plt.axis("off")
        single_plots_graph(.05, .84, 0.4, .1, [0,1], "Arm: li-re", data)
        single_plots_graph(.50, .84, 0.4, .1, [2], "Arm: komb.", data)
        single_plots_graph(.05, .69, 0.4, .1, [3], "Oberkörper", data)
        single_plots_graph(.50, .69, 0.4, .1, [4], "Final li (OK, Arm)", data)
        single_plots_graph(.05, .55, 0.4, .1, [5], "Final re (OK, Arm)", data)
        single_plots_graph(.50, .55, 0.4, .1, [6], "Final komb.", data)


        # second half of page
        plt.axes([.02, .02, .49, .5])
        plt.text(0.65,0.95, s="Prozentuale Verteilung pro Bereich")
        plt.axis("off")
        single_plots_hist(.05, .35, .4, .1, 0, "Arm: li", data)
        single_plots_hist(.50, .35, .4, .1, 2, "Arm: komb.", data)
        single_plots_hist(.05, .20, .4, .1, 3, "Oberkörper", data)
        single_plots_hist(.50, .20, .4, .1, 4, "Final li (OK, Arm)", data)
        single_plots_hist(.05, .05, .4, .1, 5, "Final re (OK, Arm)", data)
        single_plots_hist(.50, .05, .4, .1, 6, "Final komb.", data)


        pdf.savefig()  # saves the current figure into a pdf page

        # -2- plotting second page
        fig2 = plt.figure(figsize=(width, height), dpi=200)
        image = im.imread(image_path[0])
        subImg = plt.axes([.06, .4, .5, .55])
        #plt.axis("off")

        # adding a table to the picture showing statistical parameters
        rows = ["Oberarm", "Unterarm", "Hand", "Hals", "Oberkörper", "Beine"]
        cols = ["Mean", "SD", "Variance"]
        cell_text = np.random.randint(0,90, size=(6, 3)).astype(str)
        plt.table(cellText = cell_text, rowLabels = rows, colLabels = cols,
          loc='right', cellLoc = 'center', rowLoc = 'left', colLoc = 'center',
          colWidths = [0.2 for x in cols])
        plt.subplots_adjust(left=.9)

        # drawing circles onto the skeleton
        circle = pt.Circle((200, 200), radius=50, color="green", fill=True, alpha=.9)
        subImg.add_artist(circle)
        plt.imshow(image, interpolation="sinc")
        #plt.axis("off")

        # third third of page
        ampel_plot(data)

        pdf.savefig()

        plt.close()

        # We can also set the file's metadata via the PdfPages object:
        d = pdf.infodict()
        d['Title'] = 'RUlA Report'
        d['Author'] = 'Velamed GmbH'
        d['Subject'] = 'RULA Report based on myoMOTION data'
        d['Keywords'] = 'myoMOTION, RULA, Velamed'
        d['CreationDate'] = datetime.datetime.today()
        #d['ModDate'] = datetime.datetime.today()


test_array = np.random.rand(5000, 7)
name = "test"
rula_to_pdf(name, test_array)
